crates/wysiwyg/src/dom/nodes/text_node.rs

```diff
@@ -96,6 +106,68 @@ where
             false
         }
     }
+
+    /// This gets the character at the cursor, considering the
+    /// direction of travel
+    pub fn char_at_offset(
+        &self,
+        offset: usize,
+        direction: &Direction,
+    ) -> Option<char> {
+        self.data()
+            .chars()
+            .nth(direction.get_index_from_cursor(offset))
+    }
+
+    /// This gets the character type at the cursor, considering the
+    /// direction of travel
+    pub fn char_type_at_offset(
+        &self,
+        offset: usize,
+        direction: &Direction,
+    ) -> Option<CharType> {
+        let char = self.char_at_offset(offset, direction);
+        match char {
+            Some(c) => Some(get_char_type(c)),
+            None => None,
+        }
+    }
+
+    /// When moving through a node, the cursor counts as inside the node
+    /// at one end, but not the other. This function determines that.
+    pub fn offset_is_inside_node(
+        &self,
+        current_offset: usize,
+        direction: &Direction,
+    ) -> bool {
+        let node_length = self.data().len();
+        match direction {
+            Direction::Forwards => current_offset < node_length,
+            Direction::Backwards => current_offset > 0,
+        }
+    }
+
+    /// Required due to zero length text node existence
+    pub fn has_length(&self) -> bool {
+        self.data().len() != 0
+    }
+}
+
+/// Given a character, determine it's type
+fn get_char_type(c: char) -> CharType {
+    // in order to determine where a ctrl/opt + delete type operation finishes
+    // we need to distinguish between whitespace (nb no newline characters), punctuation
+    // and then everything else is treated as the same type
+    if c.is_whitespace() || c == ZWSP_char {
+        // manually add zero width space character
+        return CharType::Whitespace;
+    }
+
+    if c.is_ascii_punctuation() {
+        return CharType::Punctuation;
+    }
+
+    return CharType::Other;
```

Comment:
Maybe this is more 'Rusty'?

```suggestion
    if c.is_whitespace() || c == ZWSP_char {
        // manually add zero width space character
       CharType::Whitespace
    } else if c.is_ascii_punctuation() {
        CharType::Punctuation
    } else {
        CharType::Other
    }
```

Reply:
Done